### ISSUE TITLE
PPA builds: Generating helptags fails

### ISSUE THREAD

**Issue by fwalch** [neovim/neovim#2189]
PPA builds for all distributions/architectures fail when generating helptags (Git revision: 608d709fa0878cbc4ff7386badc81fdfcc66234a):

```
CMake Error at /build/buildd/neovim-0.0.0+git201503190142+1435+8+7~ubuntu14.10.1/cmake/GenerateHelptags.cmake:25 (message):
  Generating helptags failed: [?1049h[H[J[H[J[1;1H[1;1H - Child
  aborted
```

Full build log e.g. [here](https://launchpad.net/~neovim-ppa/+archive/ubuntu/unstable/+build/7088877).

**Comment by nanomad** [comment]
I've found some issues with helptags on the Archlinux PKGBUILD too. Either deadlocking with 100% CPU or failing.

Here's a core dump of a deadlock issue (note that it  contanains a matching binary and source distribution):

https://transfer.sh/NEOJd/neovim-helptags-highcpu.zip

**Comment by yngwin** [comment]
I'm running into the same issue on Gentoo:

    -- Generating helptags in /var/tmp/portage/app-editors/neovim-9999/image//usr/share/nvim/runtime/doc.
    CMake Error at /var/tmp/portage/app-editors/neovim-9999/work/neovim-9999/cmake/GenerateHelptags.cmake:25 (message):
      Generating helptags failed:
      ^[[?1049h^[[H^[[2J^[[?2004h^[[H^[[2J^[[1;1H^[[1;1H^[[?12l^[[?25h - Child aborted

**Comment by khalidchawtany** [comment]
Same error here on OS X 10.10.2, using @tarruda's [builtin-terminal-emulation] (https://github.com/tarruda/neovim/tree/builtin-terminal-emulation) branch.
This happened only after I updated today afternoon. I had already updated today morning and everything was OK.

```
>> make --debug=v CMAKE_BUILD_TYPE=RelWithDebInfo CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX:PATH=/usr/local/Cellar/neovim/HEAD" install

...
...

-- Generating helptags in /usr/local/Cellar/neovim/HEAD/share/nvim/runtime/doc.
CMake Error at /Library/Caches/Homebrew/neovim--git/cmake/GenerateHelptags.cmake:25 (message):
  Generating helptags failed:

  - Child aborted
Call Stack (most recent call first):
  cmake_install.cmake:511 (include)


make[1]: *** [install] Error 1
make: *** [install] Error 2

```

**Comment by Pyrohh** [comment]
Hmm, it just happened twice in a row for me on @tarruda's branch. I run `make distclean && make`, then `sudo make install`, and get the same error. The directory structure at /usr/local looks like this.

```
local/
└── share
    └── nvim

2 directories, 0 files
```

**Comment by khalidchawtany** [comment]
@Pyrohh the error happens during ```make install``` process. As a result of this, I cannot get @tarruda's branch to work.
No :term back to **tmux** :(

**Comment by tarruda** [comment]
I think I know whats wrong, but Im not near my laptop right now

If I'm correct, removing the event:* commit will fix the issue(but will temporarily reintroduce the resize error)

**Comment by Pyrohh** [comment]
@khalidchawtany does this issue still happen for you on the latest revision of #2076? It's fixed for me (well, I ran `sudo rm -rf -- /usr/local && sudo make install` a few times and it worked fine).

**Comment by tarruda** [comment]
@fwalch from the log I can't tell what may have caused the abort. Is it possible to generate a core dump and get a backtrace from it? To enable core dumps you need to call something like "ulimit -c unlimited" prior to generating the tags.

**Comment by khalidchawtany** [comment]
It is fixed. Thank you. :+1:

**Comment by fwalch** [comment]
@tarruda I don't think I can do that easily.. but the error still happens (on master), consistently for all Ubuntu distributions. I can't reproduce it locally in my Ubuntu chroot (I also never had problems on Arch Linux, btw).

**Comment by tarruda** [comment]
>  Generating helptags failed: [?1049h[H[J[H[J[1;1H[1;1H - Child

These are terminfo strings for entering altscreen, clearing the screen and moving the cursor. All are printed at startup, but why they are being displayed by the log right before an abort is a mystery to me.

>  I can't reproduce it locally in my Ubuntu chroot (I also never had problems on Arch Linux, btw).

Does the same PPA build script succeeds in your chroot?

**Comment by tarruda** [comment]
I think I know the cause of this, I just tested starting nvim with all standard streams redirected to /dev/null and it aborted. This patch fixes it:

```diff
diff --git a/src/nvim/main.c b/src/nvim/main.c
index 47bb2bc..d97344b 100644
--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -392,7 +392,8 @@ int main(int argc, char **argv)
   if (!params.headless) {
     // Stop reading from stdin, the UI layer will take over now
     input_stop_stdin();
-    ui_builtin_start();
+    ui_builtin_start((params.input_isatty || params.err_isatty)
+        && (params.output_isatty || params.err_isatty));
   }

   setmouse();  // may start using the mouse
diff --git a/src/nvim/ui.c b/src/nvim/ui.c
index e22ef66..eaa78a3 100644
--- a/src/nvim/ui.c
+++ b/src/nvim/ui.c
@@ -70,9 +70,11 @@ static int height, width;
 #define UI_CALL_MORE(method, ...) if (ui->method) ui->method(ui, __VA_ARGS__)
 #define UI_CALL_ZERO(method) if (ui->method) ui->method(ui)

-void ui_builtin_start(void)
+void ui_builtin_start(bool tty)
 {
-  tui_start();
+  if (tty) {
+    tui_start();
+  }
 }

 void ui_builtin_stop(void)
```

Can you apply it to a PPA build to see if it fixes the problem?

**Comment by fwalch** [comment]
> Does the same PPA build script succeeds in your chroot?

I'm using `bzr dailydeb` and `pbuilder` to build locally, I thought it would be more or less the same as the Launchpad builder..

Anyway, using the patch in a test PPA seems to have [worked](https://code.launchpad.net/~fwalch/+archive/ubuntu/neovim-testing/+build/7108704). So as soon as this is fix hits master, the PPA builds should work again.

**Comment by tarruda** [comment]
> Anyway, using the patch in a test PPA seems to have worked. So as soon as this is fix hits master, the PPA builds should work again.

Added the commit to #2076, will merge as soon as the travis build passes

**Comment by justinmk** [comment]
@tarruda 

```c
    ui_builtin_start((params.input_isatty || params.err_isatty)
        && (params.output_isatty || params.err_isatty));
```

is equivalent to:

```c
    ui_builtin_start(params.input_isatty || (params.output_isatty && params.err_isatty);
```

Is that intended?

**Comment by tarruda** [comment]
> Is that intended?

:+1: Pushed the simplified version

**Comment by tarruda** [comment]
That patch didn't go very well with the old tests, so I removed it fro #2076 and will deal with it later

**Comment by nanomad** [comment]
Archlinux issues may be related to the distribution itself, I've changed the package manager and it seems to work fine now. 

There's another error with a test failing randomly, but I'll open a new issue for that

**Comment by grtlr** [comment]
Could it be that this issue is related to #2248?

**Comment by tarruda** [comment]
> Could it be that this issue is related to #2248?

I think it is

Can someone confirm that the problem is fixed on master?

**Comment by grtlr** [comment]
#2248 is now fixed for me on master.

**Comment by syswow** [comment]
#2248  confirm fixed, thank you.

**Comment by fwalch** [comment]
PPA build works again, thanks @tarruda!
